File: kotlinRpnSolver/src/exceptions/AggregateException.kt
package exceptions

class AggregateException : Exception {
    val innerExceptions: List<Throwable>

    constructor(message: String, exceptions: List<Throwable>) : super(message) {
        innerExceptions = exceptions.toList() // Create a defensive copy
    }

    constructor(message: String, cause: Throwable, exceptions: List<Throwable>) : super(message, cause) {
        innerExceptions = exceptions.toList() // Create a defensive copy
    }

    override val message: String
        get() = "${super.message}\nInner Exceptions:\n" +
                innerExceptions.mapIndexed { index, ex ->
                    "  ${index + 1}. ${ex.javaClass.simpleName}: ${ex.message}"
                }.joinToString("\n")

    override fun toString(): String = "${javaClass.simpleName}: $message"

    override fun printStackTrace(s: java.io.PrintStream) {
        s.println(this)
        innerExceptions.forEachIndexed { index, ex ->
            s.println("Inner Exception #${index + 1}:")
            ex.printStackTrace(s)
        }
    }
}

File: dartRpnSolver/exceptions/AggregateException.dart
class AggregateException implements Exception {
  final List<Exception> innerExceptions;
  final String _message;
  final Exception? cause;

  AggregateException(this._message, List<Exception> exceptions, [this.cause])
      : innerExceptions = List.unmodifiable(exceptions);

  @override
  String toString() {
    final buffer = StringBuffer();
    buffer.writeln(_message);
    buffer.writeln('Inner Exceptions:');

    for (int i = 0; i < innerExceptions.length; i++) {
      final ex = innerExceptions[i];
      buffer.writeln('  ${i + 1}. ${ex.runtimeType}: ${ex.toString()}');
    }

    return buffer.toString().trimRight();
  }

  void printStackTrace() {
    print(toString());

    for (int i = 0; i < innerExceptions.length; i++) {
      print('Inner Exception #${i + 1}:');
      print(innerExceptions[i]);
    }
  }
}
